### DISCUSSION TITLE
Custom Action Help: No Model, How to Page Data, and Button Requires 2 Clicks

### DISCUSSION THREAD

**Discussion by falconsmilie** [filamentphp/filament#4648] [Help]
Hi there!

I created an **Action** i can reuse. It's purpose is to preview a resource in a template. 

Documentation for actions is light. I've been reading through the actions contained in Filament core but I'm still failing to grasp why I am having some issues. The code follows and i've commented where i am having problems;

```
use Filament\Pages\Actions\Action;

class GetPreviewAction extends Action
{

    // some setter code here for the model, explained below ...

    protected function setUp(): void
    {
        $this->action(function(): void {

            // this feels a bit dodgy ... better way to do this?
            $data = $this->getLivewire()->data;

            // $this->getModel() will always return null ... 
            // my hack is to set the model when i call the action from the create/edit record class
            $record = $this->previewModel::newModelInstance();

            $record->fill($data);

            // it always takes 2 "clicks" on the "Preview" button for the new tab to open.
            $this->url(
                fn (): string => route('preview-page', $record),
                true
            );
        });
    }
}
```

**Comment by danharrin** [comment]
Its taking 2 clicks as `$this->action()` means that the code will only be executed on the first click, and then the URL is active from the second click.

```php
protected function setUp(): void
{
    parent::setUp() // THIS SHOULD ALWAYS BE CALLED

    $this->url(
        fn (): string => route('preview-page', $this->getRecord()),
        true
    );
}
```

And then just pass the record to the action when you instantiate it:

```
GetPreviewAction::make()->record($this->record)
```

**Reply by falconsmilie** [reply]
Thanks. Appreciate your feedback.

**Reply by falconsmilie** [reply]
Hi @danharrin . Your suggested solution works great if the resource needs to be previewed directly after page load.

However, once changes are made via the create/edit form, these changes are not reflected in the record. 

Is there a way to update the Action's reference to the record, after the user changes values in the form fields? Simialr to how the form fields themelves have the `afterStateUpdated` method.

**Reply by falconsmilie** [reply]
The goal is to create a "live preview" which functions similar to [Statamic's](https://statamic.dev/live-preview#overview). Doesn't need to be as extensive, just simply opening a new tab is fine.

**Reply by mondherinfo** [reply]
Hello, did you find a method to get data from the form each time you update your form (edit & create form), if yes can you share it please ?

**Reply by falconsmilie** [reply]
Check out this; https://github.com/pboivin/filament-peek

**Reply by mondherinfo** [reply]
i did it custom action , works with all relationships for the model and with each activated language, 
thanks
